pkg/controller/cognitoidentityprovider/userpool/setup.go

```diff
@@ -302,31 +343,181 @@ func areUserPoolAddOnsEqual(spec *svcapitypes.UserPoolAddOnsType, current *svcsd
+		if err != nil {
+			return true, errors.Wrap(err, "SmsVerificationMessage and verificationMessageTemplate.SmsMessage")
+		}
+		if !fieldUpToDate {
+			return false, nil
+		}
 	}
-	return true
+
+	return true, nil
+}
+
+// conflictingFieldsHelper checks if 2 *string fields conflict and if their value isUpToDate
+func conflictingFieldsHelper(field1 *string, field2 *string, fieldAWS *string) (bool, error) {
+	// both of them nil => all fine
+	if field1 == nil && field2 == nil {
+		return true, nil
+	}
+	if field1 != nil && field2 != nil {
+		if awsclients.StringValue(field1) != awsclients.StringValue(field2) {
+			// both of them non-nil and different => means conflict
+			return true, errors.New(errConflictingFields)
+		}
+		// both of them non-nil, but same => check if value isUpToDate
+		if awsclients.StringValue(field1) != awsclients.StringValue(fieldAWS) {
+
+			return false, nil
+		}
+
+		return true, nil
+	}
+	// check which one is non-nil and if its value isUpToDate
+	if field1 != nil {
+		return awsclients.StringValue(field1) == awsclients.StringValue(fieldAWS), nil
+	}
+	return awsclients.StringValue(field2) == awsclients.StringValue(fieldAWS), nil
 }
 
 func areVerificationMessageTemplateEqual(spec *svcapitypes.VerificationMessageTemplateType, current *svcsdk.VerificationMessageTemplateType) bool {
 	if spec != nil && current != nil {
-		switch {
+		switch { // EmailMessage, EmailSubject, SmsMessage are checked for in conflictingFieldsEqual
 		case awsclients.StringValue(spec.DefaultEmailOption) != awsclients.StringValue(current.DefaultEmailOption),
-			awsclients.StringValue(spec.EmailMessage) != awsclients.StringValue(current.EmailMessage),
 			awsclients.StringValue(spec.EmailMessageByLink) != awsclients.StringValue(current.EmailMessageByLink),
-			awsclients.StringValue(spec.EmailSubject) != awsclients.StringValue(current.EmailSubject),
-			awsclients.StringValue(spec.EmailSubjectByLink) != awsclients.StringValue(current.EmailSubjectByLink),
-			awsclients.StringValue(spec.SmsMessage) != awsclients.StringValue(current.SmsMessage):
+			awsclients.StringValue(spec.EmailSubjectByLink) != awsclients.StringValue(current.EmailSubjectByLink):
 			return false
 		}
 	}
 	return true
 }
 
+func (e *hooks) areMFAConfigEqual(cr *svcapitypes.UserPool) (bool, error) {
+
+	out, err := e.client.GetUserPoolMfaConfig(&svcsdk.GetUserPoolMfaConfigInput{UserPoolId: awsclients.String(meta.GetExternalName(cr))})
+	if err != nil {
+		return true, errors.Wrap(err, errFailedGetMFARequest)
+	}
+
+	// out should not be nil, bc we set MFA to OFF in preCreate
+	if awsclients.StringValue(cr.Spec.ForProvider.MFAConfiguration) != awsclients.StringValue(out.MfaConfiguration) {
+		return false, nil
+	}
+
+	// only check MFAConfig stuff, if MFAConfiguration is ON/OPTIONAL, bc AWS doesn't allow setting stuff in SetUserPoolMfaConfig if MFA is OFF (-> e.g. Token enabled in specs with MFA OFF )
```

Comment:
Please adjust the width of your comment lines to be at most 80 characters long to improve readability.

Reply:
Adjusted comment lines in setup.go and userpool.yaml (+ added info)